kafka_utils/kafka_rolling_restart/main.py

```diff
@@ -279,6 +312,24 @@ def wait_for_stable_cluster(
         time.sleep(check_interval)
 
 
+def execute_task(checks, host):
+    """execute all the prechecks for the host
+    This trying to run(), and on getting a PrecheckFailedException executes failure().
+    It will try to run run() before exiting(incase of faiilure) to assert if
+    precheck is now valid()
+    """
+    for check in checks:
+        try:
+            check.run(host)
+        except TaskFailedException:
+            check.failure(host)
```

Comment:
The way this has been structured is every task is executed, and upon execution, there is an outcome. A user may want to do some success or failure things on a task execution. This is why I made it this way.

Also on failure I am triggering puppet upgrade, from the task function. This allows me to have less logic in main.py and move all the update else logic in these pluggable modules.